Immutable HttpFields and MetaData

gregw opened the pull request:
This is an experimental branch to see the impact of making `MetaData` immutable.  This is currently against 10, but may be for a 10.1 or later version.

gregw committed:
Immutable version of HttpFields

Preserve API and usage of HttpFields class while providing a read only interface and immutable implementation.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable version of HttpFields

Use an ArrayList in HttpFields. While slightly slower than the array, it will mostly be used as a builder pattern for an Immutable

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
First few commits are about creating an immutable `HttpFields`.  The `HttpField` class itself was already immutable.

gregw committed:
Immutable version of HttpFields

Fixed exception type.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable version of HttpFields

asImmutable method

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable MetaData

Made HttpURIU immutable with a builder pattern.
MetaData immutable and working within http module.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
@joakime @sbordet @janbartel @lorban @lachlan-roberts    I've made `HttpFields`, `HttpURI` and `MetaData` immutable and fixed things up to the http module.  It's going to take a while to get everything else fixed, but would be good to get feedback before I commit too much more effort if the tree I'm barking up is not going to work.    So if you wanted to review those three classes whilst this is a broken WIP, I'd appreciate any comments etc.  (ok not any comments... )

lorban commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java:
```diff
@@ -724,337 +466,163 @@ public HttpField remove(HttpHeader name)
+        public Iterator<HttpField> iterator()
         {
-            return _cursor + 1;
+            return new Iterator<>()
```
Equivalent to:
`return Stream.of(_fields).iterator();`

lorban commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:
```diff
@@ -50,527 +49,126 @@
+    private final String _query;
+    private final String _fragment;
+    private String _uri;
+    private String _decodedPath;
```
Why aren't the above two private?

lorban commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpFields.java:
```diff
@@ -45,12 +36,9 @@
  * <p>The cookie handling provided by this class is guided by the Servlet specification and RFC6265.
  */
-public class HttpFields implements Iterable<HttpField>
+public class HttpFields implements Iterable<HttpField>, HttpFieldList
```
This class' immutability is... surprising. I'm not fond of the internal interface TBH.

gregw replied:
We are limited in how much we can change HttpFields because it is part of some public interfaces.   This change make `HttpFields` a builder for the `HttpFields.Immutable` inner class.    It's not what I'd do if green fields.     Note that for `HttpURI` I have made it truly immutable and added `HttpURI.Builder`, which I think is a better way to go... I did that differently because that is more of an internal class.

I'm not sure what you mean by "internal interface"?  There is an external `HttpEventList` interface that is implemented by both `HttpFields` and `HttpFields.Immutable`

lorban commented on jetty-http/src/main/java/org/eclipse/jetty/http/MetaData.java:
```diff
@@ -18,35 +18,44 @@
+    private final long _contentLengthKnown;
+    private final long _contentLengthField;
+    private final Supplier<HttpFieldList> _trailerSupplier;
+    private HttpFieldList _trailers;
```
Why isn't that field immutable?

gregw replied:
Because it is not part of the immutable meta-data.  rather it is a field of the self referencing trailerSupplier that is used if a specific supplier is not provided.
I will review this once I have trailers working and will then document a lot better if I keep it... I may move to an inner class if I can convince myself it wont be less efficient.

lorban commented:
There's a lack of coherency to quickly figure out if a class is immutable or not or can be made immutable. Maybe we should introduce some form of common pattern for classes that can sometimes be mutable or immutable (like the collections' `Collection.unmodifiable()` for instance) and/or a naming convention to distinguish them?

gregw committed:
Immutable MetaData

Fixes from review

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable MetaData

Passing tests upto and including jetty-server

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
I've pushed changes that makes tests pass up to and including `jetty-server`.   So I think it is worthwhile tacking step back and reviewing the approaches, which as @lorban says, have a "lack of coherency"

Currently I've taken 3 different approaches to immutability:
1. `HttpURI` is now immutable itself and has an explicit `Builder` for mutable operations like:
```java
   HttpURI newUri = new HttpURI.Builder(oldUri).host(serverName).port(serverPort).normalize().build();
```

2. `HttpFields` has taken the opposite approach, keeping that class as a mutable implementing the current API.  It has a method `asImmutable()` that returns a `HttpFieldList`, which is a new interface that has all the read only methods of `HttpFields`.  This interface is implemented by both `HttpFields` and it's private `HttpFields.Immutable` implementation.  Because their is no builder pattern, the code to modify is a bit more verbose because there is no chaining: 
```java
   HttpFields newFields = new HttpFields(oldFieldList);
   newFields.remove(someHeader);
   newFields.put("Header", "replacement value");
   HttpFieldList newFieldList = newFields.asImmutable();
``` 

3. `MetaData`, `MetaData.Request` and `MetaData.Response` are purely immutable with no builder and only constructors.  This is somewhat of a pain in the server as http/1 parser delivers all the components in separate callbacks, so they need to be remembered before making the metadata instance when the headers are complete.   This could be a bit more elegant and readable with either a builder pattern or a mutable version of itself.

One immutable variation I have not used is having an API that is read/write, but the immutable version returns a `UnsupportedOperationException` for write methods.  This could perhaps tidy up the `HttpFields` approach but only if we were able to make `HttpFields` an interface which would break backwards compatibility.

Also I don't particularly like the Builder pattern being so explicit for HttpURI, so perhaps that could be made a bit cleaning with some utility methods so you could write:
```java
   HttpURI newUri = HttpURI.copy(oldUri).host(serverName).port(serverPort).normalize().toHttpURI();
```
Perhaps this approach could also be used to add a builder to `HttpFields`:

```java
   HttpFields.copy(oldFields).remove(someHeader).put("Header","replace value").toHttpFields();
``` 
This would make HttpFields immutable, so it would not be possible to have code like:
```java
   HttpFields fields = new HttpFields();
   fields.add("Header", "value");
   fields.add("Host", "localhost");
```
This would instead need to be written as
```java
   HttpFields fields = HttpFields.empty()
       .add("Header", "value")
       .add("Host", "localhost")
       .toHttpFields();
```

If we decide to go that way, then I think the same approach would work for `MetaData`.

I may have a fiddle with this....  but I'd really like a bit more feedback before I commit too much more effort.  @sbordet specially for h2 and the client.

lorban commented:
For your inspiration, I was thinking about using a pair of marker interfaces that would explicitly state what classes are builders and how to derive a new object from a builder-generated one.

Something like: 
```
public interface Builder<T>
{
  T build();
}

public interface Derivable<B extends Builder<? extends Derivable<?>>>
{
  B derive();
}
```
which would give us:
```
public class HttpFieldsBuilder implements Builder<HttpFields>
{
  public HttpFields build()
  {
  ...
  }
}

public class HttpFields implements Derivable<HttpFieldsBuilder>
{
  public HttpFieldsBuilder derive()
  {
  ...
  }
}

public class Sample
{
  public static void main()
  {
    HttpFields original = new HttpFieldsBuilder()
      .add("Header", "value")
      .add("Host", "localhost")
      .build();

    HttpFields copy = original.derive()
      .remove("Header")
      .put("Port", "8080")
      .build();
  }
}

```

These interfaces would have minimal added value, but would mostly serve as "marker" interfaces, explicitly letting the programmer know that using some class requires following the common builder / derive pattern.

gregw committed:
Immutable MetaData

Cleanup of HttpURI.Builder API as suggested in PR.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable MetaData

Added builder for MetaData.Request

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Merge remote-tracking branch 'origin/jetty-10.0.x' into jetty-10.0.x-ImmutableMetaData

gregw committed:
Immutable MetaData

more api fixes

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable MetaData

WIP making HttpFiels itself immutable.  Currently working up to jetty-servlet.

Need to consider if content-length really is meta data and how much and when can we trust it.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable MetaData

WIP

Need to consider if content-length really is meta data and how much and when can we trust it. Also need to consider difference between h2 and h1 authority in metadata.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
@sbordet can I get a few hours of your time early next week to consider the impact of this on the client and h2.

gregw committed:
Immutable MetaData WIP

jetty-client and jetty-servlet passing tests.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable MetaData WIP

Better align the style of immutability between `HttpFields` and `HttpURI`.
They both now have static build() and from() methods, plus Builder and Immutable implementations.
Potentially `Builder` could be renamed as `Mutable`

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable MetaData WIP

http2-server tests passed

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable MetaData WIP

http2-client tests passed

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable MetaData WIP

cleann build?

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Merge branch 'jetty-10.0.x' into jetty-10.0.x-ImmutableMetaData

gregw committed:
Immutable MetaData WIP

fix

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable MetaData WIP

more test fixes

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable MetaData WIP

Cleanups, mostly using EMPTY when appropriate.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable MetaData WIP

Cleanups, use immutable

Signed-off-by: Greg Wilkins <gregw@webtide.com>

joakime commented:
Will this eventually mean we'll have a `HttpConfiguration.getBaseResponseHttpFields()` that provides the baseline response fields for the connectors that the configuration belongs to?

(With things like `Server: Jetty 10.0.x` already present in it?)

gregw commented:
@lorban,  I've ended up kind of similar to what you suggest, but without the marker interfaces. Both HttpFields and HttpURI are like:
```java
interface HttpFoo
{
    Object getSome();
    Object getThing();
    HttpFoo asImmutable();
    Builder asMuttable();
    static Builder build() { return new Builder(); }
    static Builder build(HttpFoo copy) { return new Builder(copy); }
    static Builder build(String representation) { return new Builder(representation); }
    static HttpFoo from(String representation) { return new Builder(representation).asImmutable(); // may be optimised }

    class Builder implements HttpFoo
    {
         Object _some;
         Object _thing;
         private Builder() { ... };
         private Builder(HttpFoo copy) { ... };
         private Builder(String representation) { ... };
         
        Object getSome() { return _some; }
        Object getThing() { return _thing; }
        Builder some(Object s) { _some = s; return this; }
        Builder thing(Object t) { _thing = t; return this; }
        HttpFoo asImmutable() { return new Immutable(_some, _thing); }
        Builder asMuttable() { return this };
    }

    class Immutable implements HttpFoo
    {
         final Object _some;
         final Object _thing;
         private Immutable(Object s, Object t) { ... };
         
        Object getSome() { return _some; }
        Object getThing() { return _thing; }
        HttpFoo asImmutable() { return this; }
        Builder asMuttable() { return new Builder(this); };
    }
}

class Holder
{
    private HttpFoo _foo;
    public Holder(Http foo) { _foo = foo.asImmutable(); }
    public HttpFoo getFoo() {return _foo;}
    public void setFoo(HttpFoo foo) { _foo = foo.asImmutable(); }
}

public static void main(String... arg)
{
    HttpFoo ex1 = HttpFoo.from("ex:1");
    HttpFoo ex2 = HttpFoo.build(ex1).thing("2").asImmutable();
    HttpFoo lazy2 = HttpFoo.build(ex2).some("lazy"));

    HttpFoo.Builder build = HttpFoo.build()
        .some("some")
        .thing("thing");
    otherStuff();
    build.thing("thing!");

    Holder h1 = new Holder(build.asImmutable());  // explicit
    Holder h2 = new Holder(build);  // sufficient

    h1.setFoo(HttpFoo.build(h1.getFoo()).some("SOME"));
}
```

So this works more or less as your `derive()` and `build()` methods, except that every object has both methods.   So if you are passed a `HttpFoo` but don't know if the underlying implementation is mutable or not, then you just call `asImmutable()` and either way you get an immutable copy.   This way both caller and callee can make decisions.  You could put `HttpFoo.Immutable` in your signature if you really only wanted to be passed an immutable, or you can just have `HttpFoo` and then internally decide if you want to take an immutable copy (which might not copy if it was already immutable) or not.

I'm almost tempted to rename `Builder` to `Mutable` in that pattern,  but I do like the `build(...)` methods.   There are also possible tweaks like should it be 'HttpFoo.Immutable from(String representation)` or just `HttpFoo from(String representation)`.  I've done the former... but am leaning towards the later.

gregw commented:
> Will this eventually mean we'll have a HttpConfiguration.getBaseResponseHttpFields() that provides the baseline response fields for the connectors that the configuration belongs to? (With things like Server: Jetty 10.0.x already present in it?)

@joakime Not this PR...... but that is a really good idea anyway and could remove a few if's and inefficient settings.  Can you open an issue to do that and we can get it done for a final 10 even if this PR doesn't make it in.

Edit: we could keep the existing setters, but rather than set booleans they can modify the base HttpFields

gregw committed:
Immutable MetaData WIP

No trailers for connect

Signed-off-by: Greg Wilkins <gregw@webtide.com>

joakime commented:
> > Will this eventually mean we'll have a HttpConfiguration.getBaseResponseHttpFields() that provides the baseline response fields for the connectors that the configuration belongs to? (With things like Server: Jetty 10.0.x already present in it?)
> 
> @joakime Not this PR...... but that is a really good idea anyway and could remove a few if's and inefficient settings. Can you open an issue to do that and we can get it done for a final 10 even if this PR doesn't make it in.
> 
> Edit: we could keep the existing setters, but rather than set booleans they can modify the base HttpFields

I opened Issue #4795 for you

gregw committed:
Immutable MetaData WIP

Fix CONNECT path handling

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable MetaData WIP

fixed rewrite query handling

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable MetaData WIP

rename Builders to Muttables

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable MetaData WIP

misc cleanups

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Merge remote-tracking branch 'origin/jetty-10.0.x' into jetty-10.0.x-ImmutableMetaData

gregw commented:
I have modified the pattern a little bit:
```java
interface HttpFoo
{
    Object getSome();
    Object getThing();
    Immutable asImmutable();
    static Mutable build() { return new Builder(); }
    static Mutable build(HttpFoo copy) { return new Builder(copy); }
    static Mutable build(String representation) { return new Builder(representation); }
    static Immutable from(String representation) { return new Builder(representation).asImmutable(); // may be optimised }

    class Mutable implements HttpFoo
    {
         Object _some;
         Object _thing;
         private Mutable() { ... };
         private Mutable(HttpFoo copy) { ... };
         private Mutable(String representation) { ... };
         
        Object getSome() { return _some; }
        Object getThing() { return _thing; }
        Mutable some(Object s) { _some = s; return this; }
        Mutable thing(Object t) { _thing = t; return this; }
        Immutable asImmutable() { return new Immutable(_some, _thing); }
    }

    class Immutable implements HttpFoo
    {
         final Object _some;
         final Object _thing;
         private Immutable(Object s, Object t) { ... };
         
        Object getSome() { return _some; }
        Object getThing() { return _thing; }
        Immutable asImmutable() { return this; }
    }
}

class Holder
{
    private HttpFoo _foo;
    public Holder(Http foo) { _foo = foo.asImmutable(); }
    public HttpFoo getFoo() {return _foo;}
    public void setFoo(HttpFoo foo) { _foo = foo.asImmutable(); }
}

public static void main(String... arg)
{
    HttpFoo ex1 = HttpFoo.from("ex:1");
    HttpFoo ex2 = HttpFoo.build(ex1).thing("2").asImmutable();
    HttpFoo lazy2 = HttpFoo.build(ex2).some("lazy"));

    HttpFoo.Mutable build = HttpFoo.build()
        .some("some")
        .thing("thing");
    otherStuff();
    build.thing("thing!");

    Holder h1 = new Holder(build.asImmutable());  // explicit
    Holder h2 = new Holder(build);  // sufficient

    h1.setFoo(HttpFoo.build(h1.getFoo()).some("SOME"));
}
```

gregw commented:
@sbordet @lorban @joakime This PR is ready for review now!
As this is huge, I suggest we review in three phases:
1. Look at `HttpURI` and `HttpFields` first.   You may need to look at some of their usages to see how the changed API works, but focus primarily on the design of these classes.
2. Look at the wider usage of `HttpURI` and `HttpFields`. Check that changes made are all correct and in the best style.   I have focused primarily on non test classes, so there are some tests that have mechanical updates and may not use the best technique... but I have also manually edited hundreds of tests, so most of them should be good.
3. Finally look at the `MetaData` changes.  I'm very happy with the changes made in server request/response handling, but I would appreciate close scrutiny of HttpClient and h2 usage.  There still feels like there are too many complex constructors for these classes (although I have reduced a bit).

I'm thinking this is appropriate for 10.

gregw commented:
@lorban could you run your garbage test against this branch?

gregw committed:
Immutable MetaData WIP

Revert to using arrays due to garbage generated by streams and iterators (12% of a simple benchmark!).
Even if this garbage is an artifact of the JIT being disabled by observation, it can hide other allocations, so best to just use simple arrays!

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable MetaData WIP

More optimizations and better test coverage.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable Metadata

various cleanups

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Merge branch 'jetty-10.0.x' into jetty-10.0.x-ImmutableMetaData

joakime commented:
I'm +1 for merging into `jetty-10.0.x`

gregw committed:
Immutable MetaData WIP

More optimizations

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Merge remote-tracking branch 'origin/jetty-10.0.x' into jetty-10.0.x-ImmutableMetaData

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable MetaData WIP

review changes

Signed-off-by: Greg Wilkins <gregw@webtide.com>

sbordet commented on jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/client/HttpUpgraderOverHTTP.java:
```diff
@@ -30,13 +30,19 @@
 
 public class HttpUpgraderOverHTTP implements HttpUpgrader
 {
+    private static final PreEncodedHttpField WS_VERSIONS_FIELD = new PreEncodedHttpField(HttpHeader.SEC_WEBSOCKET_VERSION, WebSocketConstants.SPEC_VERSION_STRING);
```
Please rename this field to `WS_VERSION_FIELD` (singular).

sbordet commented on jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/client/HttpUpgraderOverHTTP2.java:
```diff
@@ -23,12 +23,14 @@
 
 public class HttpUpgraderOverHTTP2 implements HttpUpgrader
 {
+    public static final PreEncodedHttpField SPEC_VERSION_FIELD = new PreEncodedHttpField(HttpHeader.SEC_WEBSOCKET_VERSION, WebSocketConstants.SPEC_VERSION_STRING);
```
This field is the same field that exist in `HttpUpgraderOverHTTP`.
Can you please name them the same, i.e. `WS_VERSION_FIELD`?

sbordet commented on jetty-client/src/main/java/org/eclipse/jetty/client/HttpRequest.java:
```diff
@@ -353,7 +381,7 @@ public Request attribute(String name, Object value)
 
     @Override
-    public HttpFields getHeaders()
+    public HttpFields.Mutable getHeaders()
```
Should not this be just `HttpFields` as return type?

gregw replied:
Actually now that I have added the `put(HttpField)`, `add(HttpField)` etc. methods it could be an immutable return.
So currently this supports both old style (get the fields as a builder and modify) and the new style (use `add`,`put`, `set`).

We can keep it like this, go back to just old style (removing new methods) or change to immutable to just support new style.  Your thoughts?

sbordet replied:
See #4808. Let's finish this and then discuss - currently I'm leaning for `Request headers(Consumer<HttpFields.Mutable>)`

gregw replied:
I did try the consumer approach and it worked well enough, but was a little bit more verbose.   Will defer this to #4808

sbordet commented on jetty-client/src/main/java/org/eclipse/jetty/client/HttpResponse.java:
```diff
@@ -103,15 +108,15 @@ public HttpFields getHeaders()
     }
 
-    public HttpFields getTrailers()
+    public HttpFields.Mutable getTrailers()
```
Should not this be just `HttpFields` as return type?

sbordet commented on jetty-client/src/main/java/org/eclipse/jetty/client/api/Request.java:
```diff
@@ -152,6 +153,32 @@
+     * @see #header(HttpHeader, String)
+     */
+    Request put(HttpField field);
+
```
I think for now these are fine, although `remove()` is never used.
I will file another issue about reviewing the way headers can be set on a request because there is need to explicitly separate the "add" semantic from the "put" semantic.

gregw replied:
See comment above... maybe these are not necessary anymore?

sbordet commented on jetty-client/src/main/java/org/eclipse/jetty/client/util/MultiPartRequestContent.java:
```diff
@@ -145,7 +145,7 @@ public void fail(Throwable failure)
      * @param fields the headers associated with this part
      */
-    public void addFieldPart(String name, Request.Content content, HttpFields fields)
+    public void addFieldPart(String name, Request.Content content, HttpFields.Mutable fields)
```
Remove `.Mutable`.

sbordet commented on jetty-client/src/main/java/org/eclipse/jetty/client/util/MultiPartRequestContent.java:
```diff
@@ -164,7 +164,7 @@ public void addFieldPart(String name, Request.Content content, HttpFields fields
      * @param fields the headers associated with this part
      */
-    public void addFilePart(String name, String fileName, Request.Content content, HttpFields fields)
+    public void addFilePart(String name, String fileName, Request.Content content, HttpFields.Mutable fields)
```
Remove `.Mutable`.

sbordet commented on jetty-client/src/test/java/org/eclipse/jetty/client/ClientConnectionCloseTest.java:
```diff
@@ -249,8 +249,7 @@ public void handle(String target, Request baseRequest, HttpServletRequest reques
                 HttpConnectionOverHTTP connection = (HttpConnectionOverHTTP)connectionPool.getActiveConnections().iterator().next();
                 assertFalse(connection.getEndPoint().isOutputShutdown());
-            })
-            .onResponseHeaders(r -> r.getHeaders().remove(HttpHeader.CONNECTION));
```
This line was the point of the test: verify that even if the server did not send `Connection: close`, the client closes the connection.
It must be restored.

gregw replied:
So do you want to the signature of `Response.HeaderListener` to be:
```java
boolean onHeader(Response response, HttpField.Mutable field);
```
Currently I have not exposed the mutable for the response headers.

Is it expected that listeners in the client mutate the headers received or is this just for this test?

sbordet replied:
Ah good point. It is just a trick for the test. `HttpClient` detects when there is a `Connection: close` and initiates the close locally. I want to remove the `Connection: close` so that I want to verify that the TCP close by the server triggers the client to TCP close too. However, I can see applications wanting to filter out certain headers (e.g. cookies) so perhaps a `Mutable` is not that bad in this listener.

gregw replied:
For now I have just down cast in the unit test to access the mutable headers, so the test works.  I'll defer to #4808 for any other changes to the response listener APIs

sbordet commented on jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientRedirectTest.java:
```diff
@@ -70,7 +70,7 @@ public void test303(Scenario scenario) throws Exception
         assertNotNull(response);
         assertEquals(200, response.getStatus());
-        assertFalse(response.getHeaders().containsKey(HttpHeader.LOCATION.asString()));
+        assertFalse(response.getHeaders().contains(HttpHeader.LOCATION.asString()));
```
Since you have replaced `containsKey()` with `contains()`, can you also remove all the `.asString()` in this class, now that `contains()` takes a `HttpHeader`?

sbordet commented on jetty-documentation/src/main/java/embedded/client/http/HTTPClientDocs.java:
```diff
@@ -78,6 +79,11 @@
+
+    public static final PreEncodedHttpField H2C_UPGRADE_FIELD = new PreEncodedHttpField(HttpHeader.UPGRADE, "h2c");
+    public static final PreEncodedHttpField H2_SETTINGS_FIELD = new PreEncodedHttpField(HttpHeader.HTTP2_SETTINGS, "");
+    public static final PreEncodedHttpField H2_UPGRADE_CONNECTION_FIELD = new PreEncodedHttpField(HttpHeader.CONNECTION, "Upgrade, HTTP2-Settings");
```
Remove these fields, this is documentation code, we don't need to optimize.

sbordet commented on jetty-documentation/src/main/java/embedded/client/http/HTTPClientDocs.java:
```diff
@@ -798,9 +804,9 @@ public void dynamicClearText() throws Exception
-            .header(HttpHeader.CONNECTION, "Upgrade, HTTP2-Settings")
+            .add(H2C_UPGRADE_FIELD)
+            .add(H2_SETTINGS_FIELD)
+            .add(H2_UPGRADE_CONNECTION_FIELD)
```
Restore this section of the documentation as it was before, no point in optimizing it.

sbordet commented on jetty-documentation/src/main/java/embedded/client/http2/HTTP2ClientDocs.java:
```diff
@@ -172,11 +172,11 @@ public void newStreamWithData() throws Exception
         // Configure the request headers.
-        HttpFields requestHeaders = new HttpFields();
-        requestHeaders.put(HttpHeader.CONTENT_TYPE, "application/json");
+        HttpFields.Mutable requestHeaders = HttpFields.build()
```
Remove `.Mutable`.

sbordet commented on jetty-documentation/src/main/java/embedded/client/http2/HTTP2ClientDocs.java:
```diff
@@ -213,9 +213,9 @@ public void responseListener() throws Exception
-        HttpFields requestHeaders = new HttpFields();
-        requestHeaders.put(HttpHeader.USER_AGENT, "Jetty HTTP2Client {version}");
-        MetaData.Request request = new MetaData.Request("GET", new HttpURI("http://localhost:8080/path"), HttpVersion.HTTP_2, requestHeaders);
+        HttpFields.Mutable requestHeaders = HttpFields.build()
```
Remove `.Mutable`.

sbordet commented on jetty-documentation/src/main/java/embedded/client/http2/HTTP2ClientDocs.java:
```diff
@@ -358,9 +358,9 @@ public void pushReset() throws Exception
-        HttpFields requestHeaders = new HttpFields();
-        requestHeaders.put(HttpHeader.USER_AGENT, "Jetty HTTP2Client {version}");
-        MetaData.Request request = new MetaData.Request("GET", new HttpURI("http://localhost:8080/path"), HttpVersion.HTTP_2, requestHeaders);
+        HttpFields.Mutable requestHeaders = HttpFields.build()
```
Remove `.Mutable`.

sbordet commented on jetty-documentation/src/main/java/embedded/server/http/HTTPServerDocs.java:
```diff
@@ -464,9 +464,7 @@ public void handle(String target, Request jettyRequest, HttpServletRequest reque
-                    HttpURI newURI = new HttpURI(uri);
-                    String newPath = "/new_path/" + path.substring("/old_path/".length());
-                    newURI.setPath(newPath);
+                    HttpURI newURI = HttpURI.build(uri).path("/new_path/" + path.substring("/old_path/".length()));
```
Restore the local variable `newPath` for documentation clarity.

sbordet commented on jetty-fcgi/fcgi-client/src/main/java/org/eclipse/jetty/fcgi/client/http/HttpSenderOverFCGI.java:
```diff
@@ -74,13 +71,14 @@ protected void sendHeaders(HttpExchange exchange, ByteBuffer contentBuffer, bool
-        httpField = headers.remove(HttpHeader.CONTENT_TYPE);
+        httpField = headers.getField(HttpHeader.CONTENT_TYPE);
         fcgiHeaders.put(FCGI.Headers.CONTENT_TYPE, httpField == null ? "" : httpField.getValue());
+        headers.remove(EnumSet.of(HttpHeader.AUTHORIZATION, HttpHeader.CONTENT_LENGTH, HttpHeader.CONTENT_TYPE));
```
I feel better to use `remove()` field by field, rather than `getField()` + `remove(EnumSet)`.
Reason being that if there is one more/less header to treat in this way, I will never remember to add/remove it from the `EnumSet`.

gregw replied:
Actually it can be even simplified more.  No need for headers to be mutable anyway!

sbordet commented on jetty-fcgi/fcgi-server/src/main/java/org/eclipse/jetty/fcgi/server/proxy/FastCGIProxyServlet.java:
```diff
@@ -180,7 +180,7 @@ protected void sendProxyRequest(HttpServletRequest request, HttpServletResponse
 
         // If the Host header is missing, add it.
-        if (!proxyRequest.getHeaders().containsKey(HttpHeader.HOST.asString()))
+        if (!proxyRequest.getHeaders().contains(HttpHeader.HOST.asString()))
```
Remove `asString()`.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpParser.java:
```diff
@@ -72,7 +72,7 @@
  * (eg. "Connection: close"), or just header names (eg. "Connection:" ).
  * For headers who's value is not known statically (eg. Host, COOKIE) then a
- * per parser dynamic Trie of {@link HttpFields} from previous parsed messages
+ * per parser dynamic Trie of {@link HttpFields.Mutable} from previous parsed messages
```
Remove `.Mutable`.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:
```diff
@@ -18,798 +18,1064 @@
- * <code>http://user@host:port/path/info;param?query#fragment</code>
- * this class will split it into the following undecoded optional elements:<ul>
+ *
+ * Both {@link HttpFields.Mutable} and {@link HttpFields.Immutable} implementations are available
```
Don't refer to `HttpFields`.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:
```diff
@@ -18,798 +18,1064 @@
  * <p>Any parameters will be returned from {@link #getPath()}, but are excluded from the
  * return value of {@link #getDecodedPath()}.   If there are multiple parameters, the
- * {@link #getParam()} method returns only the last one.
+ * {@link #getParam()} method returns only the last one. Fragments are ignored.
```
Restore fragments.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:
```diff
@@ -18,798 +18,1064 @@
+            _decodedPath = null;
+            _param = null;
+            _query = null;
+        }
```
Please use the same order of the fields declarations, otherwise it's too difficult to say if we forgot to assign a field.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:
```diff
@@ -18,798 +18,1064 @@
+                return true;
+            if (!(o instanceof HttpURI))
+                return false;
+            return toString().equals(o.toString());
```
Don't this is a good idea to use `toString()` for equality, since we may modify `toString()` to include the hashCode and now we are screwed.
Introduce an `asString()` method or `toURI()` with a clear semantic and use those.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:
```diff
@@ -18,798 +18,1064 @@
+            if (!(o instanceof HttpURI))
+                return false;
+            return toString().equals(o.toString());
+        }
```
The implementation of `hashCode()` is missing! It must be there if `equals()` is overridden.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/HttpURI.java:
```diff
@@ -18,798 +18,1064 @@
+            _decodedPath = null;
+            _user = null;
+            return this;
+        }
```
Use same order as the field declarations.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/MetaData.java:
```diff
@@ -24,29 +24,38 @@
 public class MetaData implements Iterable<HttpField>
 {
-    private HttpVersion _httpVersion;
+    private static final Supplier<HttpFields> SELF_SUPPLIED_TRAILORS = () -> null;
```
Typo: TRAILERS.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/MetaData.java:
```diff
@@ -83,42 +84,39 @@ public HttpFields getFields()
+    public boolean mayHaveTrailers()
+    {
+        return _trailers != null;
+    }
```
Please review whether we really need this.

sbordet commented on jetty-http/src/main/java/org/eclipse/jetty/http/MetaData.java:
```diff
@@ -257,7 +221,10 @@ public ConnectRequest(HttpScheme scheme, HostPortHttpField authority, String pat
         public ConnectRequest(String scheme, HostPortHttpField authority, String path, HttpFields fields, String protocol)
         {
-            super(HttpMethod.CONNECT.asString(), scheme, authority, path, HttpVersion.HTTP_2, fields, Long.MIN_VALUE);
+            // TODO should connect requests have paths?
```
Yes, CONNECT has a scheme and a path when it's a WebSocket tunnel over HTTP/2 (RFC8441).

sbordet commented on jetty-http/src/test/java/org/eclipse/jetty/http/HttpURIParseTest.java:
```diff
@@ -179,7 +179,10 @@ public void testParseString(String input, String scheme, String host, Integer po
+
+            // We don't retain the fragment
+            if (input.indexOf('#') >= 0)
+                input = input.substring(0, input.indexOf('#'));
```
Revert when you have reverted the fragment.

sbordet commented on jetty-http2/http2-client/src/test/java/org/eclipse/jetty/http2/client/HTTP2Test.java:
```diff
@@ -811,8 +806,8 @@ public void testClientInvalidHeader() throws Exception
         Session session = newClient(new Session.Listener.Adapter());
-        HttpFields requestFields = new HttpFields();
-        requestFields.put(":custom", "special");
+        HttpFields.Mutable requestFields = HttpFields.build()
```
Remove `.Mutable`.

sbordet commented on jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/MetaDataBuilder.java:
```diff
@@ -162,14 +163,6 @@ else if (value != null)
-                            _authority = (HostPortHttpField)field;
-                        else if (value != null)
-                            _authority = new AuthorityHttpField(value);
-                    }
```
Why was this logic removed? Seems important to me.

gregw replied:
I don't believe it is necessary anymore because the `Request.setMetaData` method is able to take the authority from either the URI or the host field.
There is no need to look for a host header in a HTTP2 request when the authority field is mandatory.

We had this kind of dual approach were we were trying to make the h2 metadata look a bit like h1 meta data, but not entirely.... so the request still has to be aware of the differences.  So either we make metadata identical regardless of protocol - OR we just let it record the info that was in the message and let the client/server deal with it.   I've opted for the later.

gregw replied:
I have checked the proxy tests for h2->http/1.1 and the host header is correctly added by the HttpClient.

sbordet commented on jetty-http2/http2-http-client-transport/src/main/java/org/eclipse/jetty/http2/client/http/HttpSenderOverHTTP2.java:
```diff
@@ -64,6 +64,8 @@ protected void sendHeaders(HttpExchange exchange, ByteBuffer contentBuffer, bool
         MetaData.Request metaData;
         if (isTunnel)
         {
+            // TODO what to do with trailers ??
```
This is the CONNECT tunnel request, it does not have trailers.

sbordet commented on jetty-http2/http2-http-client-transport/src/main/java/org/eclipse/jetty/http2/client/http/HttpSenderOverHTTP2.java:
```diff
@@ -64,6 +64,8 @@ protected void sendHeaders(HttpExchange exchange, ByteBuffer contentBuffer, bool
         if (isTunnel)
         {
+            // TODO what to do with trailers ??
+            // TODO should connect requests have paths???
```
CONNECT requests may have scheme and path for WebSocket over HTTP/2.

sbordet commented on jetty-http2/http2-server/src/main/java/org/eclipse/jetty/http2/server/HttpChannelOverHTTP2.java:
```diff
@@ -110,21 +110,9 @@ public Runnable onRequest(HeadersFrame frame)
-                    fields.put("host", authority);
-                }
-            }
-
```
This code was removed, but seems important that it is there.

gregw replied:
I don't think so.  H2 requests do not need a host header and the `Request.setMetaData` is able to extract authority from either URI or field as appropriate.

gregw replied:
I have checked the proxy tests for h2->http/1.1 and the host header is correctly added by the HttpClient.

sbordet commented on jetty-rewrite/src/main/java/org/eclipse/jetty/rewrite/handler/RewritePatternRule.java:
```diff
@@ -90,20 +91,9 @@ public String apply(String target, HttpServletRequest request, HttpServletRespon
-            request.setQueryString(queryString);
-        }
+        HttpURI baseURI = request.getHttpURI();
+        String query = URIUtil.addQueries(baseURI.getQuery(),_query);
```
Add space before `_query`.

gregw replied:
doh! checkstyle has stopped working for me! Again!

sbordet commented on jetty-rewrite/src/test/java/org/eclipse/jetty/rewrite/handler/ForwardedSchemeHeaderRuleTest.java:
```diff
@@ -73,13 +72,13 @@ public void testHeaderValue() throws Exception
         assertEquals("https", _request.getScheme());
 
-        _request.setScheme("other");
+        _request.setHttpURI(HttpURI.build(_request.getRequestURI()).scheme("other").asImmutable());
```
Is `asImmutable()` needed here?

sbordet commented on jetty-rewrite/src/test/java/org/eclipse/jetty/rewrite/handler/ForwardedSchemeHeaderRuleTest.java:
```diff
@@ -73,13 +72,13 @@ public void testHeaderValue() throws Exception
         assertEquals("other", _request.getScheme());
 
-        _request.setScheme(null);
+        _request.setHttpURI(HttpURI.build(_request.getRequestURI()).scheme((String)null).asImmutable());
```
Is `asImmutable()` needed here?

sbordet commented on jetty-rewrite/src/test/java/org/eclipse/jetty/rewrite/handler/RewriteHandlerTest.java:
```diff
@@ -83,7 +84,7 @@ public void test() throws Exception
         _handler.setRewriteRequestURI(true);
         _handler.setRewritePathInfo(true);
-        _request.setURIPathQuery("/xxx/bar");
+        _request.setHttpURI(HttpURI.build(_request.getHttpURI(), "/xxx/bar").asImmutable());
```
Remove all `asImmutable()` from this test class?

sbordet commented on jetty-rewrite/src/test/java/org/eclipse/jetty/rewrite/handler/RewritePatternRuleTest.java:
```diff
@@ -87,8 +88,7 @@ public void testRequestWithQueryString() throws IOException
         String queryString = "request=parameter";
-        _request.setURIPathQuery("/old/context");
-        _request.setQueryString(queryString);
+        _request.setHttpURI(HttpURI.build(_request.getHttpURI(), "/old/context", null, queryString).asImmutable());
```
Remove all `asImmutable()` from this test class?

sbordet commented on jetty-rewrite/src/test/java/org/eclipse/jetty/rewrite/handler/RewriteRegexRuleTest.java:
```diff
@@ -62,12 +63,12 @@ public void testRequestUriEnabled(Scenario scenario) throws Exception
 
         reset();
-        _request.setURIPathQuery(null);
+        _request.setHttpURI(HttpURI.build(_request.getHttpURI()).asImmutable());
```
Remove all `asImmutable()` from this test class?

sbordet commented on jetty-rewrite/src/test/java/org/eclipse/jetty/rewrite/handler/ValidUrlRuleTest.java:
```diff
@@ -43,7 +44,7 @@ public void init() throws Exception
     {
         _rule.setCode("404");
-        _request.setURIPathQuery("/valid/uri.html");
+        _request.setHttpURI(HttpURI.build(_request.getHttpURI(), "/valid/uri.html").asImmutable());
```
Remove all `asImmutable()` from this test class?

sbordet commented on jetty-rewrite/src/test/java/org/eclipse/jetty/rewrite/handler/VirtualHostRuleContainerTest.java:
```diff
@@ -49,7 +50,7 @@ public void init() throws Exception
 
         start(false);
-        _request.setURIPathQuery("/cheese/bar");
+        _request.setHttpURI(HttpURI.build(_request.getHttpURI(), "/cheese/bar").asImmutable());
```
Remove all `asImmutable()` from this test class?

sbordet commented on jetty-security/src/test/java/org/eclipse/jetty/security/DataConstraintsTest.java:
```diff
@@ -71,7 +72,7 @@ public void startServer()
             public void customize(Connector connector, HttpConfiguration channelConfig, Request request)
             {
-                request.setScheme(HttpScheme.HTTPS.asString());
+                request.setHttpURI(HttpURI.build(request.getHttpURI()).scheme(HttpScheme.HTTPS).asImmutable());
```
Remove all `asImmutable()` from this test class?

sbordet commented on jetty-security/src/test/java/org/eclipse/jetty/security/authentication/SpnegoAuthenticatorTest.java:
```diff
@@ -113,11 +112,11 @@ public void flush() throws IOException {}
-        MetaData.Request metadata = new MetaData.Request(httpFields);
-        metadata.setURI(new HttpURI("http://localhost"));
+
+        MetaData.Request metadata = new MetaData.Request(null, HttpURI.build("http://localhost").asImmutable(), null,
```
Remove all `asImmutable()` from this test class?

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java:
```diff
@@ -720,7 +715,7 @@ public void onTrailers(HttpFields trailers)
         if (LOG.isDebugEnabled())
             LOG.debug("onTrailers {} {}", this, trailers);
-        _trailers = trailers;
+        _request.getMetaData().setTrailers(trailers.asImmutable());
```
Remove `asImmutable()`?

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -1163,15 +1119,19 @@ public HttpVersion getHttpVersion()
 
     public String getQueryEncoding()
+    {
+        return _queryEncoding == null ? null : _queryEncoding.toString();
```
Should be `_queryEncoding.name()`, not `toString()`.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java:
```diff
@@ -488,10 +488,8 @@ protected boolean passConditionalHeaders(HttpServletRequest request, HttpServlet
                 // Find multiple fields by iteration as an optimization 
-                HttpFields fields = ((Request)request).getHttpFields();
-                for (int i = fields.size(); i-- > 0; )
+                for (HttpField field :  ((Request)request).getHttpFields())
```
Space after colon.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java:
```diff
@@ -2142,16 +2140,25 @@ public RequestDispatcher getRequestDispatcher(String uriInContext)
+                    builder.path(URIUtil.addPaths(contextPath,builder.getPath()));
+                    uri = builder.asImmutable();
+                    pathInfo = uri.getDecodedPath().substring(contextPath.length());
+                }
```
The `if` branch is equal to the initial part of the `else` branch, so perhaps you can remove the duplication?

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java:
```diff
@@ -56,7 +56,7 @@
  * {@link Server#setErrorHandler(ErrorHandler)}.
  * It is called by the HttpResponse.sendError method to write an error page via {@link #handle(String, Request, HttpServletRequest, HttpServletResponse)}
- * or via {@link #badMessageError(int, String, HttpFields)} for bad requests for which a dispatch cannot be done.
+ * or via {@link #badMessageError(int, String, HttpFields.Mutable)} for bad requests for which a dispatch cannot be done.
```
Remove `.Mutable`.

gregw replied:
That one actually is a needed Mutable.

sbordet commented on jetty-server/src/main/java/org/eclipse/jetty/server/handler/ErrorHandler.java:
```diff
@@ -521,10 +521,12 @@ protected void writeErrorPageStacks(HttpServletRequest request, Writer writer)
      * @return The content as a ByteBuffer, or null for no body.
      */
-    public ByteBuffer badMessageError(int status, String reason, HttpFields fields)
+    public ByteBuffer badMessageError(int status, String reason, HttpFields.Mutable fields)
```
Remove `.Mutable`.

gregw replied:
that one is needed

sbordet commented on jetty-server/src/test/java/org/eclipse/jetty/server/ResponseTest.java:
```diff
@@ -861,7 +861,7 @@ public void testEncodeRedirect()
         Response response = getResponse();
         Request request = response.getHttpChannel().getRequest();
-        request.setAuthority("myhost", 8888);
+        request.setHttpURI(HttpURI.build(request.getHttpURI()).host("myhost").port(8888).asImmutable());
```
Remove `.asImmutable()`.

sbordet commented on jetty-server/src/test/java/org/eclipse/jetty/server/ResponseTest.java:
```diff
@@ -949,10 +950,12 @@ public void testSendRedirect()
-                        request.setAuthority(host, port);
-                    request.setURIPathQuery("/path/info;param;jsessionid=12345?query=0&more=1#target");
+                        builder.host(host).port(port);
+                    request.setHttpURI(builder.asImmutable());
```
Remove `.asImmutable()`.

sbordet commented on jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/core/extensions/PerMessageDeflaterBufferSizeTest.java:
```diff
@@ -106,7 +106,7 @@ public void testClientDeflateBufferSize() throws Exception
         upgradeRequest.addExtensions("permessage-deflate; @deflate_buffer_size=" + deflateBufferSize);
 
-        CompletableFuture<HttpFields> futureRequestHeaders = new CompletableFuture<>();
+        CompletableFuture<HttpFields.Mutable> futureRequestHeaders = new CompletableFuture<>();
```
Likely no need for `.Mutable` everywhere in this class.

sbordet commented on jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/impl/JettyClientUpgradeRequest.java:
```diff
@@ -49,14 +49,14 @@ public JettyClientUpgradeRequest(WebSocketCoreClient coreClient, UpgradeRequest
             if (cookies != null)
             {
-                HttpFields headers = getHeaders();
+                HttpFields.Mutable headers = getHeaders();
```
This is the same as the variable `fields` declared above, so remove the duplication?

sbordet commented on tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/ProxyWithDynamicTransportTest.java:
```diff
@@ -442,7 +442,7 @@ protected void close(Throwable failure)
         Session session = sessionPromise.get(5, TimeUnit.SECONDS);
         String serverAddress = "localhost:" + serverConnector.getLocalPort();
-        MetaData.ConnectRequest connect = new MetaData.ConnectRequest(HttpScheme.HTTP, new AuthorityHttpField(serverAddress), null, new HttpFields(), null);
+        MetaData.ConnectRequest connect = new MetaData.ConnectRequest(HttpScheme.HTTP, new AuthorityHttpField(serverAddress), null, HttpFields.build(), null);
```
Use `HttpFields.EMPTY`.

sbordet commented on tests/test-http-client-transport/src/test/java/org/eclipse/jetty/http/client/ProxyWithDynamicTransportTest.java:
```diff
@@ -527,7 +527,7 @@ protected void close(Throwable failure)
         Session session = sessionPromise.get(5, TimeUnit.SECONDS);
         String serverAddress = "localhost:" + serverConnector.getLocalPort();
-        MetaData.ConnectRequest connect = new MetaData.ConnectRequest(HttpScheme.HTTP, new AuthorityHttpField(serverAddress), null, new HttpFields(), null);
+        MetaData.ConnectRequest connect = new MetaData.ConnectRequest(HttpScheme.HTTP, new AuthorityHttpField(serverAddress), null, HttpFields.build(), null);
```
Use `HttpFields.EMPTY`.

gregw committed:
Immutable MetaData WIP

changes after review:
 + less usage of Mutable
 + more usage of EMPTY
 + restored fragment handling

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable MetaData WIP

changes after review:
 + less usage of Mutable
 + less usage of asImmutable

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable MetaData WIP

changes after review:
 + less usage of Mutable

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Immutable MetaData

changes after review:
 + better handling of URI in ContextHandler

Signed-off-by: Greg Wilkins <gregw@webtide.com>

sbordet commented on jetty-documentation/src/main/java/embedded/client/http2/HTTP2ClientDocs.java:
```diff
@@ -266,9 +266,9 @@ public void reset() throws Exception
-        requestHeaders.put(HttpHeader.USER_AGENT, "Jetty HTTP2Client {version}");
-        MetaData.Request request = new MetaData.Request("GET", new HttpURI("http://localhost:8080/path"), HttpVersion.HTTP_2, requestHeaders);
+        HttpFields requestHeaders = HttpFields.build()
+            .add(HttpHeader.USER_AGENT, "Jetty HTTP2Client {version}");
```
Please use `put()` since it's not a multiple occurrences header.

sbordet commented on jetty-documentation/src/main/java/embedded/client/http2/HTTP2ClientDocs.java:
```diff
@@ -296,9 +296,9 @@ public void push() throws Exception
-        requestHeaders.put(HttpHeader.USER_AGENT, "Jetty HTTP2Client {version}");
-        MetaData.Request request = new MetaData.Request("GET", new HttpURI("http://localhost:8080/path"), HttpVersion.HTTP_2, requestHeaders);
+        HttpFields requestHeaders = HttpFields.build()
+            .add(HttpHeader.USER_AGENT, "Jetty HTTP2Client {version}");
```
Please use `put()` since it's not a multiple occurrences header.

gregw replied:
really?  We can see that it doesn't exist already and add is a lot faster that put... I guess not that much faster on an empty fields. ok

gregw committed:
Immutable MetaData

changes after review:
 + downcast in test to access mutable response headers.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

sbordet commented on jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/client/HttpUpgraderOverHTTP2.java:
```diff
@@ -23,12 +23,14 @@
 
 public class HttpUpgraderOverHTTP2 implements HttpUpgrader
 {
+    public static final PreEncodedHttpField WS_VERSIONS_FIELD = new PreEncodedHttpField(HttpHeader.SEC_WEBSOCKET_VERSION, WebSocketConstants.SPEC_VERSION_STRING);
```
Please rename to singular `WS_VERSION_FIELD`.

sbordet commented on jetty-websocket/websocket-jetty-client/src/main/java/org/eclipse/jetty/websocket/client/impl/JettyClientUpgradeRequest.java:
```diff
@@ -49,19 +49,16 @@ public JettyClientUpgradeRequest(WebSocketCoreClient coreClient, UpgradeRequest
-                {
-                    headers.add(HttpHeader.COOKIE, cookie.toString());
-                }
+                    fields.add(HttpHeader.COOKIE, cookie.toString());
```
Checkstyle will complain, as `for` always requires braces.

gregw replied:
Checkstyle is dead in intellij for me.  I've tried re-importing, but it just refuses to work?!?!?!?!

sbordet reviewed (changes requested):
Almost there.

gregw committed:
Immutable MetaData

changes after review:
 + use put instead of add for one time headers

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
@sbordet are we there yet?

sbordet commented on jetty-websocket/websocket-core/src/main/java/org/eclipse/jetty/websocket/core/client/HttpUpgraderOverHTTP2.java:
```diff
@@ -23,12 +23,14 @@
 
 public class HttpUpgraderOverHTTP2 implements HttpUpgrader
 {
+    public static final PreEncodedHttpField WS_VERSION_FIELD = new PreEncodedHttpField(HttpHeader.SEC_WEBSOCKET_VERSION, WebSocketConstants.SPEC_VERSION_STRING);
```
Make it `private`.

gregw replied:
done

gregw committed:
private

Signed-off-by: Greg Wilkins <gregw@webtide.com>

joakime reviewed (approved):
I'm happy with this too.

joakime commented:
Last build failed.

joakime commented:
Can you merge in the current `jetty-10.0.x` and let CI build cleanly once?

gregw committed:
Merge remote-tracking branch 'origin/jetty-10.0.x' into jetty-10.0.x-ImmutableMetaData

gregw committed:
Merge branch 'jetty-10.0.x' into jetty-10.0.x-ImmutableMetaData
